### PR TITLE
Get task status generic method

### PR DESCRIPTION
**Pull request by danielui**
Example usage
 
```Python 
session = create_vManageSession(ip_address,admin_username,password,port=port)
devices = DevicesAPI(session).devices
vsmart_device = [dev for dev in devices if dev.personality == Personality.VSMART][0]

reboot_action = RebootAction(session,devices)
reboot_action.execute()

# Keep asking for reboot status until it's not in exit_statuses (Failure or Success) or timeout is not achieved (3000s)
task = wait_for_completed(session,reboot_action.action_id,3000)
if task.status == OperationStatus.SUCCESS.value:
    #do something
else:
    #do something else
```

### CODE REVIEW

**Review by tzietkowski — DISMISSED** [review]

File: vmngclient/api/task_status_api.py [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+class TaskStatus:
+    """API class to check task status"""
+
+    def __init__(self, session: vManageSession):
+        self.session = session
+        self.status: str = ''
+        self.status_id: str = ''
+        self.activity: List[str] = []
+
+    def wait_for_completed(
```
Comment (cicharka) [inline vmngclient/api/task_status_api.py:22]: Maybe docstring for this method would be a nice place for example usage?

Also, any docstring for this method should be provided here as it is the main method we are going to use from this code.

**Review by JimOverholt — COMMENTED** [review]
```python
    exit_statuses = [OperationStatus.SUCCESS.value,
                          OperationStatus.FAILURE.value]
    exit_statuses_ids = [OperationStatusId.SUCCESS.value,
                              OperationStatusId.FAILURE.value] 
```
Why do we need to set this, as this is the default pass case?
Can this be made default, and only set if needed?

File: vmngclient/api/task_status_api.py (hunk 1 above)
Comment (danielui) [reply vmngclient/api/task_status_api.py:22]: Moved usage example to class docstring

File: vmngclient/utils/operation_status.py [hunk 2]
```diff
@@ -6,3 +6,10 @@ class OperationStatus(Enum):
     FAILURE = 'Failure'
     SUCCESS = 'Success'
     SCHEDULED = 'Scheduled'
+
```
Comment (sbasan) [inline vmngclient/utils/operation_status.py:9]: why not create one enum with string Tuple (or preferably dataclass)

File: vmngclient/api/task_status_api.py [hunk 3]
```diff
@@ -0,0 +1,123 @@
...
+        self,
+        sleep_seconds: int,
+        timeout_seconds: int,
+        action_id: str,
+        exit_statuses: List[OperationStatus] = [
```
Comment (sbasan) [inline vmngclient/api/task_status_api.py:44]: can be simplified when one enum will be used with string pairs (preferably a dataclass)

File: vmngclient/api/task_status_api.py [hunk 4]
```diff
@@ -0,0 +1,123 @@
...
+            """
+
+            if (self.status in exit_statuses) and (self.status_id in exit_statuses_ids):
+                if not activity_text or activity_text in self.activity:
+                    return False
+            return True
+
+        def _log_exception(self) -> None:
+            logger.error("Operation status not achieved in given time")
+            return None
```
Comment (sbasan) [inline vmngclient/api/task_status_api.py:95]: return statement is unnecessary when using -> None annotation

File: vmngclient/api/task_status_api.py [hunk 5]
```diff
@@ -0,0 +1,123 @@
...
+            OperationStatus.SUCCESS,
+            OperationStatus.FAILURE,
+        ],
+        exit_statuses_ids: List[OperationStatusId] = [
+            OperationStatusId.SUCCESS,
+            OperationStatusId.FAILURE,
+        ],
+        activity_text: str = '',
+        action_url: str = '/dataservice/device/action/status/',
+    ) -> "TaskStatus":
```
Comment (sbasan) [inline vmngclient/api/task_status_api.py:54]: this method returns boolean not TaskStatus instance

File: vmngclient/api/task_status_api.py [hunk 6]
```diff
@@ -0,0 +1,123 @@
...
+
+        # Keep asking for reboot status until it's not in exit_statuses (Failure or Success)
+        # or timeout is not achieved (3000s)
+        TaskStatus(session).wait_for_completed(5,3000,exit_statuses,exit_statuses_ids,reboot_action.action_id)
+
+    """
+
+    def __init__(self, session: vManageSession) -> None:
+        self.session = session
+        self.status: str = ''
```
Comment (sbasan) [inline vmngclient/api/task_status_api.py:35]: lifetime of variables self.status, self.status_id, self.activity is same as in wait_for_completed() method - so should be their scope

**Review by sbasan — CHANGES_REQUESTED** [review]

File: vmngclient/api/task_status_api.py (hunk 5 above)
Comment (danielui) [reply vmngclient/api/task_status_api.py:54]: Nope, it returns TaskStatus instance. It returned bool in previous version.

File: vmngclient/api/task_status_api.py (hunk 6 above)
Comment (danielui) [reply vmngclient/api/task_status_api.py:35]: One of PEP8 rule, forces us to declare all instance attributes in __init__.

File: vmngclient/api/task_status_api.py (hunk 4 above)
Comment (danielui) [reply vmngclient/api/task_status_api.py]: Fixed

File: vmngclient/api/task_status_api.py (hunk 6 above)
Comment (sbasan) [reply vmngclient/api/task_status_api.py:35]: the point was - these should not be instance variables

File: vmngclient/api/task_status_api.py (hunk 5 above)
Comment (tehAgitto) [reply vmngclient/api/task_status_api.py:53]:
```python
"""
Returns:
    bool: True if condition is met
"""
```

File: vmngclient/api/task_status_api.py [hunk 7]
```diff
@@ -0,0 +1,122 @@
...
+        self.status_id: str = ''
+        self.activity: List[str] = []
+
+    def wait_for_completed(
+        self,
+        sleep_seconds: int,
```
Comment (JimOverholt) [inline vmngclient/api/task_status_api.py:41]:
```python
    def wait_for_completed(
        self,
        action_id: str,
        sleep_seconds: int = 5,
        timeout_seconds: int = 300,
```
I think sleep_seconds should also be optional, default to 5

Comment (danielui) [reply vmngclient/api/task_status_api.py]: Done

File: vmngclient/api/task_status_api.py [hunk 8]
```diff
@@ -0,0 +1,122 @@
+import logging
+from typing import List, cast
+
+from tenacity import retry, retry_if_result, stop_after_attempt, wait_fixed  # type: ignore
+
+from vmngclient.session import vManageSession
+from vmngclient.utils.creation_tools import get_logger_name
+from vmngclient.utils.operation_status import OperationStatus, OperationStatusId
+
+logger = logging.getLogger(get_logger_name(__name__))
```
Comment (JimOverholt) [inline vmngclient/api/task_status_api.py:10]:
```suggestion
from vmngclient.utils.operation_status import OperationStatus, OperationStatusId

logger = logging.getLogger(__name__)
```

**Review by sbasan — APPROVED** [review]
lgtm

**Comment by JimOverholt** [comment]
Needs an example usage.

**Review by JimOverholt — CHANGES_REQUESTED** [review]
Needs an example usage.  
```python
url = "/dataservice/tenantbackup/import"
files = {"file": (file.name, open(str(file), "rb"))}
response = self.session.post(url, data={}, files=files)
return wait_for_completed(self.session, response.json()["processId"])


response = self.session.get_json("/dataservice/tenantbackup/export")
status= wait_for_completed(self.session, response["processId"])
string = re.search("""file location: (.*)""", status.activity[-1])
assert string, "File location not found."
return string.group(1)
```
Here are 2 snippets from https://github.com/CiscoDevNet/vManage-client/pull/40

File: vmngclient/api/task_status_api.py [hunk 9]
```diff
@@ -0,0 +1,101 @@
...
+        action_id (str): inspected action id
+        timeout_seconds (int): After this time, function will stop requesting action status
+        sleep_seconds (int): interval between action status requests
+        exit_statuses (Union[List[OperationStatus], str]): actions statuses that cause stop requesting action status
+        exit_statuses_ids (Union[List[OperationStatusId], str]): actions statuses ids
+            that cause stop requesting action status id
+        activity_text (str): activity text
+
+    Returns:
+        "TaskStatus":
```
Comment (tomaszwarda) [inline vmngclient/api/task_status_api.py:50]:
```suggestion
        teask (TaskStatus):
```

**Review by tomaszwarda — APPROVED** [review]

**Review by sbasan — APPROVED** [review]
lgtm

**Review by JimOverholt — DISMISSED** [review]

**Review by sbasan — APPROVED** [review]
lgtm

**Pull request merged**